### ISSUE TITLE
new Event Codeunit 5402 OnBeforeQtyRndPrecision

### ISSUE THREAD

**Issue by EasyStep2Lean** [Microsoft/ALAppExtensions#815]
**Rounding precision for Quantities at 5 digit still valid in D365 BC?**

Since the beginning of days even at Navi365a or even before we are limited in the rounding precision to 5 digits for the calculation between Quantity and Quantity (Base). I wonder if this is still a limit. A quite common scenario is when I have the following setup:

**Example:** 
An Item (Colour Pigments)  is purchased in Purchase UOM = PALLET, where the Loadcarrier contains 12 bags of 1 KG each. The Item is consumed in the warehouse/production in Gram's so the Base Unit Of Measure is set to Grams and the Put-Away UOM also in Gram. Receiving the material as 12 Lotnumber of each 1 KG in the Warehouse with full Lottracking can be easily posted. No we create a return order to the Vendor a Breakbulk will occur, Picking UOM=Grams converted to UOM=PALLET to the Shipment Bin. Due to rouding of only 5 digits I have as a result only 0.99999 Pallet where I need to return a full pallet.
(I can think of a hundreds of other examples mm -> M2 = 6 digits.....)

This said Dynamics NAV 2018 and D365 BC contains a nice function in Codeunit 5402 Unit Of Measure Management returning the standard 5-digit rounding precision. Which can be easily extended with a Event allowing us partners to increase the standard Rouding Precision when neccesary. 

![roundingprecisionuommgt](https://user-images.githubusercontent.com/17320737/51029072-95db5300-1595-11e9-8de4-48b843be5f64.png)

However the old code is loaded with fixed rounding like in the picture

![round-vs-call](https://user-images.githubusercontent.com/17320737/51029059-85c37380-1595-11e9-8efb-129ccc3e6986.png)

My suggestion replace all fixed ROUND(......, 0.00001); to UOMMgt.RoundQty(........); Let the default return still 5 digits and let the partners decide to increase this using the New Event. Posible add another 2 fields in Table General Ledger Setup Quantity Rounding Precision  (DECIMAL) and Quantity Decimal Places (TEXT)

Again just for Quantities, Amounts are fine for me.

Anyone an other suggestion?

**Comment by JesperSchulz** [comment]
Thanks for sharing your idea with us. We're always looking for better ways to do things. However, this isn't the right place to share it. We don't believe this is a extensibility issue - consider 2 extensions both want to set different roundings - which one wins? It would be better to control rounding as an app setting! You'll have better luck getting it implemented like that and noticed by other partners and users if you post it to [Business Central - Ideas](http://aka.ms/bcideas).

This GitHub repository handles issues with extending published apps, and extensibility in general, for the latest release of Business Central.

Thanks again for your great suggestion. We're looking forward to seeing it in our ideas page.

**Comment by EasyStep2Lean** [comment]
Well Jesper, It would be quit impossible to realize this using Extentions. So My first question would be Can we have the requested Event? But that also require you to convert all hardcoded roundings to point to this function. I can also put this down as a bug throught support channels if you prefer, since this is resulting into errors in code.

I already created a list of the related objects in D365 BC CU02 (See Excel)

I decided to use the following coding replacement above the call to function UOMMgt.RoundQuantity.

// << EWMSW113.02.01 Return rounding precision throught call in Unit Of Measure Management
//EXIT(ROUND(Qty * "Qty. per Unit of Measure",0.00001));
EXIT(ROUND(Qty * "Qty. per Unit of Measure",UOMMgt.QtyRndPrecision));
// >>
[Increase Rounding Precision Objects.xlsx](https://github.com/Microsoft/ALAppExtensions/files/2764260/Increase.Rounding.Precision.Objects.xlsx)
